[docs] Config Plugins: update code blocks, align with writing styleguide (#20974)
expo/expo: docs/pages/guides/config-plugins.mdx

```

- You may update the exact version of `expo` to build against a specific version.

> - For simple config plugins that depend on core, stable APIs, such as a plugin that only modifies **AndroidManifest.xml** ot **Info.plist**, you can use a loose dependency such as in the example above.

Reviewer: ```suggestion
- For simple config plugins that depend on core, stable APIs, such as a plugin that only modifies **AndroidManifest.xml** or **Info.plist**, you can use a loose dependency such as in the example above.
```

- You may also want to install [`expo-module-scripts`](https://github.com/expo/expo/blob/main/packages/expo-module-scripts/README.md) as a development dependency, but it's not required.

### Importing the config plugins package

The `expo/config-plugins` and `expo/config` packages are re-exported from the `expo` package.

```js
const { /* @hide ...*//* @end */ } = require('expo/config-plugins');
const { /* @hide ...*//* @end */ } = require('expo/config');
```

Importing through the `expo` package ensures that you are using the version of the `expo/config-plugins` and `expo/config` packages that are depended on by the `expo` package.

> If you do not import the package through the `expo` re-export in this way, you may accidentally be importing an incompatible version
> (depending on the implementation details of module hoisting in the package manager used by the developer consuming the module) or be unable to import the module at all
> (if using "plug and play" features of a package manager like Yarn Berry or pnpm).

Reviewer: ```suggestion
If you do not import the package through the `expo` re-export in this way, you may accidentally be importing an incompatible version
(depending on the implementation details of module hoisting in the package manager used by the developer consuming the module) or be unable to import the module at all
(if using "plug and play" features of a package manager such as Yarn Berry or pnpm).
```

### Modifying the AndroidManifest.xml

> Packages should attempt to use the built-in **AndroidManifest.xml** [merging system](https://android-doc.github.io/tools/building/manifest-merge.html)

Reviewer: ```suggestion
Packages should attempt to use the built-in **AndroidManifest.xml** [merging system](https://developer.android.com/studio/build/manage-manifests)
```

The link above doesn't work anymore. We should replace it with the official Android docs.

Author: Great catch, thank you! 🙏

### Modifying the iOS Podfile

> The iOS Podfile is the config file for CocoaPods, the dependency manager on iOS. Think of Podfile like package.json but for iOS.

Reviewer: ```suggestion
The iOS **Podfile** is the config file for CocoaPods, the dependency manager on iOS. It is similar to **package.json** for iOS.
```

> The Podfile is a ruby file (application code), which means you **cannot** safely modify it from Expo config plugins, and should opt towards another approach, like Expo Autolinking hooks (citation needed).

Reviewer: ```suggestion
The **Podfile** is a ruby file (application code), which means you **cannot** safely modify it from Expo config plugins and should opt for another approach, such as [Expo Autolinking](modules/autolinking) hooks.
```
I think we can remove the "citation needed" part and link Autolinking docs here.

### Plugin development best practices

- **Instructions in your README**: If the plugin is tied to a React Native module, then you should document manual setup instructions for the package.
  If anything goes wrong with the plugin, users should still be able to manually add the package to their project.
  Doing this often helps you to find ways to reduce the setup, which can lead to a simpler plugin.
  - Document the available properties for the plugin, and specify if the plugin works without props.
  - If you can make your plugin work after running prebuild multiple times, that’s a big plus! It can improve the developer experience to be able to run `expo prebuild` without the `--clean` flag to sync changes.

> - **Naming conventions**: Use `withFeatureName` if cross-platform. If the plugin is platform specific, use a camel case naming with the platform right after “with”. Ex; `withIosSplash`, `withAndroidSplash`.

Reviewer: ```suggestion
- **Naming conventions**: Use `withFeatureName` if cross-platform. If the plugin is platform specific, use a camel case naming with the platform right after “with”. For example, `withAndroidSplash`, `withIosSplash`.
```

  Some features are included for historical reasons, like the ability to automatically copy and link [Google services files](https://github.com/expo/expo-cli/blob/3a0ef962a27525a0fe4b7e5567fb7b3fb18ec786/packages/config-plugins/src/ios/Google.ts#L15) defined in the Expo config.
  If there is overlap, then maybe recommend the user uses the built-in types to keep your plugin as simple as possible.
- **Split up plugins by platform**: For example — `withIosSplash`, `withAndroidSplash`. This makes using the `--platform` flag in `expo prebuild` a bit easier to follow in `EXPO_DEBUG` mode.
- **Unit test your plugin**: Write Jest tests for complex modifications. If your plugin requires access to the filesystem,
  use a mock system (we strongly recommend [`memfs`][memfs]), you can see examples of this in the [`expo-notifications`](https://github.com/expo/expo/blob/fc3fb2e81ad3a62332fa1ba6956c1df1c3186464/packages/expo-notifications/plugin/src/__tests__/withNotificationsAndroid-test.ts#L34) plugin tests.

>   - Notice the root [**/**mocks\*\*\*\*](https://github.com/expo/expo/tree/main/packages/expo-notifications/plugin/__mocks__) folder and [**plugin/jest.config.js**](https://github.com/expo/expo/tree/main/packages/expo-notifications/plugin/jest.config.js).

Reviewer: ```suggestion
  - Notice the root [**/**mocks\*\*\*\*](https://github.com/expo/expo/tree/main/packages/expo-notifications/plugin/__mocks__) folder and [**plugin/jest.config.js**](https://github.com/expo/expo/tree/main/packages/expo-notifications/plugin/jest.config.js).
```

The "**/**mocks\*\*\*\*" isn't rendering correctly in HTML. I thought we can use `&lowerbar;&lowerbar;mocks&lowerbar;&lowerbar;` but Prettier in VSCode changes that to `\_\mocks\_\_`, which is fine I guess, since that pattern is also used in other paces.

TLDR; let's change this to `**\_\mocks\_\_**`.

Author: Unfortunately, `**\_\mocks\_\_**` will also brake and result in bold text since `**` wrap, but I have pushed the version which should render correctly and hopefully be fine with Prettier.

Reviewer: Thank you 🙏

- A TypeScript plugin is always better than a JavaScript plugin. Check out the [`expo-module-script` plugin][ems-plugin] tooling for more info.
- Do not modify the `sdkVersion` via a config plugin, this can break commands like `expo install` and cause other unexpected issues.

### Versioning

By default, `expo prebuild` runs transformations on a [source template][source-template] associated with the Expo SDK version that a project is using.
The SDK version is defined in the **app.json** or inferred from the installed version of `expo` that the project has.

When Expo SDK upgrades to a new version of React Native for instance, the template may change significantly to account for changes in React Native or new releases of Android or iOS.

If your plugin is mostly using [static modifications](#static-modification) then it will work well across versions.
If it's using a regular expression to transform application code, then you'll definitely want to document which Expo SDK version your plugin is intended for.

> Expo releases a new version quarterly (every 3 months), and there is a [beta period][expo-beta-docs] where you can test if your plugin works with the new version before it's released.

Reviewer: ```suggestion
Expo releases a new version quarterly (every 3 months), and there is a [beta period](https://github.com/expo/expo/blob/main/guides/releasing/Release%20Workflow.md#stage-5---beta-release) where you can test if your plugin works with the new version before it's released.
```

The link for "expo-beta-docs" docs doesn't work. I updated it here directly but feel free to move it down as how this link was referenced originally or we can delete that reference and stick with this pattern.

This can be used to quickly debug the results of [static modifications](#static-modification) without needing to run prebuild.
You can interact with introspection live, by using the [preview feature](https://github.com/expo/vscode-expo#expo-preview-modifier) of `vscode-expo`.

You can try introspection by running `expo config --type introspect` in a project.

Introspection only supports a subset of modifiers:

- `android.manifest`
- `android.gradleProperties`
- `android.strings`
- `android.colors`
- `android.colorsNight`
- `android.styles`
- `ios.infoPlist`
- `ios.entitlements`
- `ios.expoPlist`
- `ios.podfileProperties`

> Introspection only works on safe modifiers (static files like JSON, XML, plist, properties), with the exception of `ios.xcodeproj` which often requires file system changes, making it non idempotent.

Introspection works by creating custom base mods that work like the default base mods, except they don't write the `modResults` to disk at the end.
Instead of persisting, they save the results to the Expo config under `_internal.modResults`, followed by the name of the mod

> i.e. the `ios.infoPlist` mod saves to `_internal.modResults.ios.infoPlist: {}`.

Reviewer: ```suggestion
such as the `ios.infoPlist` mod saves to `_internal.modResults.ios.infoPlist: {}`.
```
